src/d3a/models/strategy/external_strategies/storage.py

```diff
@@ -122,39 +133,68 @@ def _delete_offer_impl(self, arguments, response_channel):
 
     def _offer(self, payload):
         transaction_id = self._get_transaction_id(payload)
+        required_args = {'price', 'energy', 'transaction_id'}
+        allowed_args = required_args.union({'replace_existing'})
+
         offer_response_channel = f'{self.channel_prefix}/response/offer'
         if not check_for_connected_and_reply(self.redis, offer_response_channel,
                                              self.connected):
             return
         try:
             arguments = json.loads(payload["data"])
-            assert set(arguments.keys()) == {'price', 'energy', 'transaction_id'}
+
+            # Check that all required arguments have been provided
+            assert all(arg in arguments.keys() for arg in required_args)
+            # Check that every provided argument is allowed
+            assert all(arg in allowed_args for arg in arguments.keys())
+
             arguments['seller'] = self.device.name
             arguments['seller_origin'] = self.device.name
         except Exception as e:
             logging.error(f"Incorrect offer request. Payload {payload}. Exception {str(e)}.")
             self.redis.publish_json(
                 offer_response_channel,
                 {"command": "offer",
-                 "error": "Incorrect offer request. Available parameters: (price, energy).",
+                 "error": (
+                     "Incorrect offer request. "
+                     "Available parameters: ('price', 'energy', 'replace_existing')."),
                  "transaction_id": transaction_id})
         else:
             self.pending_requests.append(
                 IncomingRequest("offer", arguments, offer_response_channel))
 
+    def can_offer_be_posted(self, time_slot, **offer_arguments):
+        """Check that the energy being offered is <= than the energy available to be sold."""
+
+        replace_existing = offer_arguments.get('replace_existing', True)
+        if replace_existing:
+            # Do not consider previously offered energy, since those offers would be deleted
+            return (
+                offer_arguments['energy'] <=
+                self.state.energy_to_sell_dict[time_slot] +
+                self.state.offered_sell_kWh[time_slot])
+        else:
+            return (
+                offer_arguments['energy'] <=
+                self.state.energy_to_sell_dict[time_slot])
+
```

Comment:
so you are overwriting the bastrategy function here for the storage, right?
I suggest to rename this function to something like `can_storage_offer_be_posted`, because this is specific to the storage. Also because of the following point.
I realise that we dont check for the `has_battery_reached_max_power` in the external storage strategy (have a look into Storage.sell_energy and Storage.buy_energy.
This should be added in here. However, As this PR is getting long, we could add it in a separate to-be-added task.  FYI @spyrostz @faizan2590

Reply:
Yes, as you say this is specific to the Storage external strategy. Actually, I named it in the same way as the one in `BaseStrategy` exactly to override that method, since that method would cause an exception if called from a `StorageExternalStrategy`.
I have the feeling that the relation between mixins and strategies is a bit problematic, or at least I sometimes confuse the two. In theory, I would assume that a class that uses a mixin should be able to call all of its methods (or required to implement its own version), but this is not the case for `StorageExternalStrategy`.

Please let me know if you still think I should proceed and rename `can_offer_be_posted` to `can_storage_offer_be_posted`.

Reply:
Since in this case `can_offer_be_posted ` is already rewritten inside `strategy/external_strategies/storage.py`, to me its  actually not really needed to explicitly name it `can_storage_offer_be_posted `

Reply:
> I realise that we don't check for the has_battery_reached_max_power in the external storage strategy (have a look into Storage.sell_energy and Storage.buy_energy.

Sorry missed that part earlier. Yes, I agree `has_battery_reached_max_power ` has to be checked too, better make a separate enhancement ticket as suggested.

Reply:
Sorry for jumping in here, but I disagree, as far as I see `energy_to_sell_dict` is already clamped by `_battery_energy_per_slot` (which is the energy that the battery can consume / produce by its max power) , have a look [here](https://github.com/gridsingularity/d3a/blob/master/src/d3a/models/state.py#L321). Therefore I think that the calculation is ok and no further checks need to be done.

Reply:
Yes, you are of course right, and we call `clamp_energy_to_sell_kWh` and `clamp_energy_to_buy_kWh` every time and offer or bid is posted. So all good. I actually was surprised when i thought that this is not being taken care of.
But we went through a lot of iterations about that already indeed.
SO, all good. ignore me. :-) Thanks a lot!

Reply:
Also, the renaming is not needed, as Faizan correctly mentioned.